manpages/ivstools.1
.TH IVSTOOLS 1 "December 2018" "Version 1.5.2"

.SH NAME
ivstools - extract IVs from a pcap file or merges several .ivs files into one
.SH SYNOPSIS
.B ivstools
--convert <pcap file> <ivs output file>
.B ivstools
--merge <ivs file 1> <ivs file 2> .. <output file>
.SH DESCRIPTION
.BI ivstools
is a tool designed to extract ivs (initialization vectors) from a pcap dump to an ivs file and it can also merge several ivs (initialization vectors) files into one..
.SH EXAMPLE
.B ivstools
--convert wep_dump.cap out.ivs
.br
.B ivstools
--merge myivs1.ivs myivs2.ivs myivs3.ivs allivs.ivs
.SH AUTHOR
This manual page was written by Adam Cecile <gandalf@le-vert.net> for the Debian system (but may be used by others).
Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 2 or any later version published by the Free Software Foundation
On Debian systems, the complete text of the GNU General Public License can be found in /usr/share/common-licenses/GPL.
.SH SEE ALSO
.br
.B airbase-ng(8)
.br
.B aireplay-ng(8)
.br
.B airmon-ng(8)
.br
.B airodump-ng(8)
.br
.B airodump-ng-oui-update(8)
.br
.B airserv-ng(8)
.br
.B airtun-ng(8)
.br
.B besside-ng(8)
.br
.B easside-ng(8)
.br
.B tkiptun-ng(8)
.br
.B wesside-ng(8)
.br
.B aircrack-ng(1)
.br
.B airdecap-ng(1)
.br
.B airdecloak-ng(1)
.br
.B airolib-ng(1)
.br
.B besside-ng-crawler(1)
.br
.B buddy-ng(1)
.br
.B kstats(1)
.br
.B makeivs-ng(1)
.br
.B packetforge-ng(1)
.br
.B wpaclean(1)
.br
.B airventriloquist(8)


CHANGELOG
Changelog for 2.0.9
-fixed client and clientless attacks chopchop and frag packet injection not generating IVs because -h MAC was not set
-force frag, chopchop and aircrack-ng xterms to stay on screen after successfull command so you can see results
-cleaned the code some more
-cleaning functions needs more tweaks
-atheros users need to use options 9 (monitor mode) before other commands to create-destroys correct VAPS
-fixed "unary operator expected" msg (harmless bug) (not fully fixed because of "" generating errors in logic tests)
-tweaked 2 steps attacks (frag, chopchop) to go to menu to inject after successfull first step
-added option to jump to injection menu (added this to the main menu)
-fixed AP encryption detection for cracking step (bug was introduced by "")
-lowered adapter speed to 1M for wesside-ng attacks
-added -hold to mdk attack to allow users to see results before the windows closes

Changelog for 2.0.8 (27/01/2008)
-added wesside-ng airoscript style usage
-added interface type detection
-tweaked menu
-first attempt at making airoscript work with different adapters (add more card support)(BIG THX TO GREEN-FREQ!!!)
-atheros device are now supported
-added mdk3
-added back to main menu options
-fixed chopchop attack (BIG THX TO GREEN-FREQ!!!)
-tweaked FAKE_MAC function for atheros

Changelog for 2.0.7 BETA1 (21/01/2007)
-changed capture from IVs to full date (ivs to cap)
-updated script for new implementation of fragmentation attack
-other minor modifications and tweaks

Changelog for PRE-1.8.2-BETA1 (30/11/2006)
-fixed all commands leading to : "head: Binary: invalid number of lines" (MAJOR THX TO ANDREA)
-fixed all commands where SSID was truncated (MAJOR THX TO PILOTSNIPES)

Changelog for 1.8.1 (03/11/2006)
-bug fix release
-fixed some problem in the SSID detection function (MAJOR THX TO andreahaku)
-cleaned a few more typos and bugs
-documentation completed a bit

Changelog for 1.8 (17/10/2006)
-dropped support for aircrack-ng 0.6.1 and older
-will only work with patched version of aircrack-ng 0.6.2 (patches can be found in the forum under the programming section)
-specifique channels hopping (for example channel 1,6,11 or 1-3,5-7,9,11)
-Access point filtering during attack (airodump will only list attacked Access point)
-new SSID detection, works with ALL ssid (supports all character and spacing possible)
-new Access point selection menu that list all AP with encryption type (eliminates the WEP? problem)
-option 2) now get you directly to a full list of AP
-chopchop attack and fragmentation attack implement
-WPA dictionnary attack implemented
-changed menu 3,4,5 based on detection of AP encryption
-added option to jump to associated client list from ap selection
-color change support for xterm
-Windows size and placement fully changeable from variables settings
-changed filenames from SSID to MAC because of possible issues if weird SSID (to eliminate security menace (if SSID is something like "rm -rf /*" etc...)
-added option an option to run "1) scan 2) select 3) attack" one after the other
-added some comments in the script to help users
-added greetings message (can be disabled setting WELCOME="1" to "0")

Changelog for 1.7RC7 (24/09/2006)
-added some resilience to the configure function now works properly with hidden ssid's and rausb devices 

Changelog for 1.7RC6 (19/09/2006)
-Code cleanup
-added deauth command to solointeractiveattack function
-changed base variables (fudge factor & inject rate)

Changelog for 1.7RC5 (14/09/2006)
-Fixed minor cosmetic issue caused by carriage return char when ssid is missing (15/9/06)
-Fixed the weird character match issue!! edit at will!!!
-added an abitrary variable ($acv) to the aircrackversioncheck function which can be used in future to keep the script backwards compatible (had to guess at what version 0.6.0 of aircrack-ng displays when run on it's own? and version 0.5.so might not work??)
-used new variable in crack function to check version and run correct crack command for the aricrack version installed fixing the bug on the B|T forum.

Changelog for 1.7RC4 (13/09/2006)
-corrected minor speellling issues ;o) 
-added aircrack-ng version detection function currently all it does is display aircrack version in the menu function (should be useful when new functionality is added to aircrack to allow backwards compatibility)
-added solointeractiveattack function (designed to run just the airreplay interactive attack function only for use with an attack that has stalled i.e. data has stopped increasing, the user can try this for another injection thread! (might not actually work?? needs more testing does spawn and seems to work!)

src/ivstools.c
/*
  *  IVS Tools - Convert or merge IVs
  *
  *  Copyright (C) 2006-2018 Thomas d'Otreppe <tdotreppe@aircrack-ng.org>
  *  Copyright (C) 2004, 2005  Christophe Devine (pcap2ivs and mergeivs)
  *
  *  This program is free software; you can redistribute it and/or modify
  *  it under the terms of the GNU General Public License as published by
  *  the Free Software Foundation; either version 2 of the License, or
  *  (at your option) any later version.
  *
  *  This program is distributed in the hope that it will be useful,
  *  but WITHOUT ANY WARRANTY; without even the implied warranty of
  *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  *  GNU General Public License for more details.
  *
  *  You should have received a copy of the GNU General Public License
  *  along with this program; if not, write to the Free Software
  *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  *
 *
 *  In addition, as a special exception, the copyright holders give
 *  permission to link the code of portions of this program with the
 *  OpenSSL library under certain conditions as described in each
 *  individual source file, and distribute linked combinations
 *  including the two.
 *  You must obey the GNU General Public License in all respects
 *  for all of the code used other than OpenSSL. *  If you modify
 *  file(s) with this exception, you may extend this exception to your
 *  version of the file(s), but you are not obligated to do so. *  If you
 *  do not wish to do so, delete this exception statement from your
 *  version. *  If you delete this exception statement from all source
 *  files in the program, then also delete it here.
 */

#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include "version.h"
#include "crypto.h"
#include "pcap.h"
#include "uniqueiv.h"
#include "aircrack-osdep/byteorder.h"
#include "aircrack-util/common.h"
#include "eapol.h"

#include "aircrack-util/console.h"

#define FAILURE -1
#define IVS 1
#define WPA 2
#define ESSID 3

/* linked list of detected access points */

struct AP_info
{
	struct AP_info * prev; /* prev. AP in list         */
	struct AP_info * next; /* next  AP in list         */

	int ssid_length; /* length of ssid           */

	unsigned char bssid[6]; /* the access point's MAC   */
	unsigned char essid[256]; /* ascii network identifier */

	unsigned char ** uiv_root; /* unique iv root structure */
	/* if wep-encrypted network */

	int wpa_stored; /* wpa stored in ivs file?   */
	int essid_stored; /* essid stored in ivs file? */
};

/* linked list of detected clients */

struct ST_info
{
	struct ST_info * prev; /* the prev client in list   */
	struct ST_info * next; /* the next client in list   */
	struct AP_info * base; /* AP this client belongs to */
	unsigned char stmac[6]; /* the client's MAC address  */
	struct WPA_hdsk wpa; /* WPA handshake data        */
};

/* bunch of global stuff */

struct globals
{
	struct AP_info *ap_1st, *ap_end;
	struct ST_info *st_1st, *st_end;

	unsigned char prev_bssid[6];
	FILE * f_ivs; /* output ivs file      */
} G;

static void usage(int what)
{
	char * version_info
		= getVersion("ivsTools", _MAJ, _MIN, _SUB_MIN, _REVISION, _BETA, _RC);
	printf("\n  %s - (C) 2006-2018 Thomas d\'Otreppe\n"
		   "  https://www.aircrack-ng.org\n"
		   "\n   usage: ",
		   version_info);
	free(version_info);
	if (what == 0 || what == 1)
		printf("ivstools --convert <pcap file> <ivs output file>\n"
			   "        Extract ivs from a pcap file\n");
	if (what == 0) printf("       ");
	if (what == 0 || what == 2)
		printf("ivstools --merge <ivs file 1> <ivs file 2> .. <output file>\n"
			   "        Merge ivs files\n");
}

static int merge(int argc, char * argv[])
{
	int i, n, unused;
	unsigned long nbw;
	unsigned char buffer[1024];
	FILE *f_in, *f_out;
	struct ivs2_filehdr fivs2;

	if (argc < 5)
	{
		usage(2);
		return (1);
	}

	// Check filenames are not empty
	for (i = 2; i < argc - 1; ++i)
	{
		if (argv[i][0] == 0)
		{
			printf("Filename #%d is empty, aborting\n", i - 1);
			return (1);
		}
	}

	printf("Creating %s\n", argv[argc - 1]);

	if ((f_out = fopen(argv[argc - 1], "wb+")) == NULL)
	{
		perror("fopen failed");
		return (1);
	}

	nbw = 0;

	for (i = 2; i < argc - 1; ++i)
	{
		printf("Opening %s\n", argv[i]);

		if ((f_in = fopen(argv[i], "rb")) == NULL)
		{
			fclose(f_out);
			perror("fopen failed");
			return (1);
		}

		if (fread(buffer, 1, 4, f_in) != 4)
		{
			fclose(f_out);
			fclose(f_in);
			perror("fread file header failed");
			return (1);
		}

		if (memcmp(buffer, IVSONLY_MAGIC, 4) == 0)
		{
			fclose(f_out);
			fclose(f_in);
			printf("%s is an old .ivs file\n", argv[i]);
			return (1);
		}

		if (memcmp(buffer, IVS2_MAGIC, 4) != 0)
		{
			fclose(f_out);
			fclose(f_in);
			printf("%s is not an .%s file\n", argv[i], IVS2_EXTENSION);
			return (1);
		}

		if (fread(&fivs2, 1, sizeof(struct ivs2_filehdr), f_in)
			!= (size_t) sizeof(struct ivs2_filehdr))
		{
			fclose(f_out);
			fclose(f_in);
			perror("fread file header failed");
			return (1);
		}

		if (fivs2.version > IVS2_VERSION)
		{
			fclose(f_out);
			fclose(f_in);
			printf("Error, wrong %s version: %d. Supported up to version %d.\n",
				   IVS2_EXTENSION,
				   fivs2.version,
				   IVS2_VERSION);
			return (1);
		}

		if (i == 2)
		{
			unused = fwrite(buffer, 1, 4, f_out);
			unused = fwrite(&fivs2, 1, sizeof(struct ivs2_filehdr), f_out);
		}

		while ((n = fread(buffer, 1, 1024, f_in)) > 0)
		{
			nbw += n;
			unused = fwrite(buffer, 1, n, f_out);
			printf("%lu bytes written\r", nbw);
		}

		fclose(f_in);

		printf("\n");
	}

	fclose(f_out);

	return (0);
}

static int dump_add_packet(unsigned char * h80211, unsigned caplen)
{
	int i, n, seq, dlen, clen;
	unsigned z;
	struct ivs2_pkthdr ivs2;
	unsigned char * p;
	unsigned char bssid[6];
	unsigned char stmac[6];
	unsigned char clear[2048];
	int weight[16];
	int num_xor, o;

	struct AP_info * ap_cur = NULL;
	struct ST_info * st_cur = NULL;
	struct AP_info * ap_prv = NULL;
	struct ST_info * st_prv = NULL;

	/* skip packets smaller than a 802.11 header */

	if (caplen < 24) return FAILURE;

	/* skip (uninteresting) control frames */

	if ((h80211[0] & 0x0C) == 0x04) return FAILURE;

	/* grab the sequence number */
	seq = ((h80211[22] >> 4) + (h80211[23] << 4));

	/* locate the access point's MAC address */

	switch (h80211[1] & 3)
	{
		case 0:
			memcpy(bssid, h80211 + 16, 6);
			break;
		case 1:
			memcpy(bssid, h80211 + 4, 6);
			break;
		case 2:
			memcpy(bssid, h80211 + 10, 6);
			break;
		default:
			memcpy(bssid, h80211 + 10, 6);
			break;
	}

	/* update our chained list of access points */

	ap_cur = G.ap_1st;
	ap_prv = NULL;

	while (ap_cur != NULL)
	{
		if (!memcmp(ap_cur->bssid, bssid, 6)) break;

		ap_prv = ap_cur;
		ap_cur = ap_cur->next;
	}

	/* if it's a new access point, add it */

	if (ap_cur == NULL)
	{
		if (!(ap_cur = (struct AP_info *) malloc(sizeof(struct AP_info))))
		{
			perror("malloc failed");
			return FAILURE;
		}

		memset(ap_cur, 0, sizeof(struct AP_info));

		if (G.ap_1st == NULL)
			G.ap_1st = ap_cur;
		else
			ap_prv->next = ap_cur;

		memcpy(ap_cur->bssid, bssid, 6);

		ap_cur->prev = ap_prv;

		ap_cur->uiv_root = uniqueiv_init();

		G.ap_end = ap_cur;

		ap_cur->ssid_length = 0;
		ap_cur->wpa_stored = 0;
		ap_cur->essid_stored = 0;
	}

	/* find wpa handshake */
	if (h80211[0] == 0x10)
	{
		/* reset the WPA handshake state */

		if (st_cur != NULL && st_cur->wpa.state != 0xFF) st_cur->wpa.state = 0;
		//        printf("initial auth %d\n", ap_cur->wpa_state);
	}

	/* locate the station MAC in the 802.11 header */

	switch (h80211[1] & 3)
	{
		case 0:

			/* if management, check that SA != BSSID */

			if (memcmp(h80211 + 10, bssid, 6) == 0) goto skip_station;

			memcpy(stmac, h80211 + 10, 6);
			break;

		case 1:

			/* ToDS packet, must come from a client */

			memcpy(stmac, h80211 + 10, 6);
			break;

		case 2:

			/* FromDS packet, reject broadcast MACs */

			if (h80211[4] != 0) goto skip_station;
			memcpy(stmac, h80211 + 4, 6);
			break;

		default:
			goto skip_station;
	}

	/* update our chained list of wireless stations */

	st_cur = G.st_1st;
	st_prv = NULL;

	while (st_cur != NULL)
	{
		if (!memcmp(st_cur->stmac, stmac, 6)) break;

		st_prv = st_cur;
		st_cur = st_cur->next;
	}

	/* if it's a new client, add it */

	if (st_cur == NULL)
	{
		if (!(st_cur = (struct ST_info *) malloc(sizeof(struct ST_info))))
		{
			perror("malloc failed");
			return FAILURE;
		}

		memset(st_cur, 0, sizeof(struct ST_info));

		if (G.st_1st == NULL)
			G.st_1st = st_cur;
		else
			st_prv->next = st_cur;

		memcpy(st_cur->stmac, stmac, 6);

		st_cur->prev = st_prv;

		G.st_end = st_cur;
	}

	if (st_cur->base == NULL || memcmp(ap_cur->bssid, BROADCAST, 6) != 0)
		st_cur->base = ap_cur;

skip_station:

	/* packet parsing: Beacon or Probe Response */

	if (h80211[0] == 0x80 || h80211[0] == 0x50)
	{
		p = h80211 + 36;

		while (p < h80211 + caplen)
		{
			if (p + 2 + p[1] > h80211 + caplen) break;

			if (p[0] == 0x00) ap_cur->ssid_length = p[1];

			if (p[0] == 0x00 && p[1] > 0 && p[2] != '\0'
				&& (p[1] > 1 || p[2] != ' '))
			{
				/* found a non-cloaked ESSID */

				//                n = ( p[1] > 32 ) ? 32 : p[1];
				n = p[1];

				memset(ap_cur->essid, 0, 256);
				memcpy(ap_cur->essid, p + 2, n);

				if (G.f_ivs != NULL && !ap_cur->essid_stored)
				{
					memset(&ivs2, '\x00', sizeof(struct ivs2_pkthdr));
					ivs2.flags |= IVS2_ESSID;
					ivs2.len += ap_cur->ssid_length;

					if (memcmp(G.prev_bssid, ap_cur->bssid, 6) != 0)
					{
						ivs2.flags |= IVS2_BSSID;
						ivs2.len += 6;
						memcpy(G.prev_bssid, ap_cur->bssid, 6);
					}

					/* write header */
					if (fwrite(&ivs2, 1, sizeof(struct ivs2_pkthdr), G.f_ivs)
						!= (size_t) sizeof(struct ivs2_pkthdr))
					{
						perror("fwrite(IV header) failed");
						return (1);
					}

					/* write BSSID */
					if (ivs2.flags & IVS2_BSSID)
					{
						if (fwrite(ap_cur->bssid, 1, 6, G.f_ivs) != (size_t) 6)
						{
							perror("fwrite(IV bssid) failed");
							return (1);
						}
					}

					/* write essid */
					if (fwrite(ap_cur->essid, 1, ap_cur->ssid_length, G.f_ivs)
						!= (size_t) ap_cur->ssid_length)
					{
						perror("fwrite(IV essid) failed");
						return (1);
					}

					ap_cur->essid_stored = 1;
					return ESSID;
				}

				for (i = 0; i < n; i++)
					if ((ap_cur->essid[i] > 0 && ap_cur->essid[i] < 32)
						|| (ap_cur->essid[i] > 126 && ap_cur->essid[i] < 160))
						ap_cur->essid[i] = '.';
			}

			p += 2 + p[1];
		}
	}

	/* packet parsing: Association Request */

	if (h80211[0] == 0x00 && caplen > 28)
	{
		p = h80211 + 28;

		while (p < h80211 + caplen)
		{
			if (p + 2 + p[1] > h80211 + caplen) break;

			if (p[0] == 0x00 && p[1] > 0 && p[2] != '\0'
				&& (p[1] > 1 || p[2] != ' '))
			{
				/* found a non-cloaked ESSID */

				n = (p[1] > 32) ? 32 : p[1];

				memset(ap_cur->essid, 0, 33);
				memcpy(ap_cur->essid, p + 2, n);

				if (G.f_ivs != NULL && !ap_cur->essid_stored)
				{
					memset(&ivs2, '\x00', sizeof(struct ivs2_pkthdr));
					ivs2.flags |= IVS2_ESSID;
					ivs2.len += ap_cur->ssid_length;

					if (memcmp(G.prev_bssid, ap_cur->bssid, 6) != 0)
					{
						ivs2.flags |= IVS2_BSSID;
						ivs2.len += 6;
						memcpy(G.prev_bssid, ap_cur->bssid, 6);
					}

					/* write header */
					if (fwrite(&ivs2, 1, sizeof(struct ivs2_pkthdr), G.f_ivs)
						!= (size_t) sizeof(struct ivs2_pkthdr))
					{
						perror("fwrite(IV header) failed");
						return (1);
					}

					/* write BSSID */
					if (ivs2.flags & IVS2_BSSID)
					{
						if (fwrite(ap_cur->bssid, 1, 6, G.f_ivs) != (size_t) 6)
						{
							perror("fwrite(IV bssid) failed");
							return (1);
						}
					}

					/* write essid */
					if (fwrite(ap_cur->essid, 1, ap_cur->ssid_length, G.f_ivs)
						!= (size_t) ap_cur->ssid_length)
					{
						perror("fwrite(IV essid) failed");
						return (1);
					}

					ap_cur->essid_stored = 1;
					return ESSID;
				}

				for (i = 0; i < n; i++)
					if (ap_cur->essid[i] < 32
						|| (ap_cur->essid[i] > 126 && ap_cur->essid[i] < 160))
						ap_cur->essid[i] = '.';
			}

			p += 2 + p[1];
		}
	}

	/* packet parsing: some data */

	if ((h80211[0] & 0x0C) == 0x08)
	{
		/* check the SNAP header to see if data is encrypted */

		z = ((h80211[1] & 3) != 3) ? 24 : 30;

		if (z + 26 > caplen) return FAILURE;

		if (z + 10 > caplen) return FAILURE;

		// check if WEP bit set and extended iv
		if ((h80211[1] & 0x40) != 0 && (h80211[z + 3] & 0x20) == 0)
		{
			/* WEP: check if we've already seen this IV */

			if (!uniqueiv_check(ap_cur->uiv_root, &h80211[z]))
			{
				/* first time seen IVs */

				if (G.f_ivs != NULL)
				{
					memset(&ivs2, '\x00', sizeof(struct ivs2_pkthdr));
					ivs2.flags = 0;
					ivs2.len = 0;

					dlen = caplen - 24 - 4 - 4; // original data len
					if (dlen > 2048) dlen = 2048;
					// get cleartext + len + 4(iv+idx)
					num_xor = known_clear(clear, &clen, weight, h80211, dlen);
					if (num_xor == 1)
					{
						ivs2.flags |= IVS2_XOR;
						ivs2.len += clen + 4;
						/* reveal keystream (plain^encrypted) */
						for (n = 0; n < (ivs2.len - 4); n++)
						{
							clear[n] = (clear[n] ^ h80211[z + 4 + n]) & 0xFF;
						}
						// clear is now the keystream
					}
					else
					{
						// do it again to get it 2 bytes higher
						num_xor = known_clear(
							clear + 2, &clen, weight, h80211, dlen);
						ivs2.flags |= IVS2_PTW;
						// len = 4(iv+idx) + 1(num of keystreams) + 1(len per
						// keystream) + 32*num_xor + 16*sizeof(int)(weight[16])
						ivs2.len += 4 + 1 + 1 + 32 * num_xor + 16 * sizeof(int);
						clear[0] = num_xor;
						clear[1] = clen;
						/* reveal keystream (plain^encrypted) */
						for (o = 0; o < num_xor; o++)
						{
							for (n = 0; n < (ivs2.len - 4); n++)
							{
								clear[2 + n + o * 32] = (clear[2 + n + o * 32]
														 ^ h80211[z + 4 + n])
														& 0xFF;
							}
						}
						memcpy(clear + 4 + 1 + 1 + 32 * num_xor,
							   weight,
							   16 * sizeof(int));
						// clear is now the keystream
					}

					if (memcmp(G.prev_bssid, ap_cur->bssid, 6) != 0)
					{
						ivs2.flags |= IVS2_BSSID;
						ivs2.len += 6;
						memcpy(G.prev_bssid, ap_cur->bssid, 6);
					}

					if (fwrite(&ivs2, 1, sizeof(struct ivs2_pkthdr), G.f_ivs)
						!= (size_t) sizeof(struct ivs2_pkthdr))
					{
						perror("fwrite(IV header) failed");
						return (1);
					}

					if (ivs2.flags & IVS2_BSSID)
					{
						if (fwrite(ap_cur->bssid, 1, 6, G.f_ivs) != (size_t) 6)
						{
							perror("fwrite(IV bssid) failed");
							return (1);
						}
						ivs2.len -= 6;
					}

					if (fwrite(h80211 + z, 1, 4, G.f_ivs) != (size_t) 4)
					{
						perror("fwrite(IV iv+idx) failed");
						return (1);
					}
					ivs2.len -= 4;

					if (fwrite(clear, 1, ivs2.len, G.f_ivs)
						!= (size_t) ivs2.len)
					{
						perror("fwrite(IV keystream) failed");
						return (1);
					}
				}

				uniqueiv_mark(ap_cur->uiv_root, &h80211[z]);
				return IVS;
			}
		}

		z = ((h80211[1] & 3) != 3) ? 24 : 30;

		if (z + 26 > caplen) return FAILURE;

		z += 6; // skip LLC header

		/* check ethertype == EAPOL */
		if (h80211[z] == 0x88 && h80211[z + 1] == 0x8E
			&& (h80211[1] & 0x40) != 0x40)
		{
			z += 2; // skip ethertype

			if (st_cur == NULL) return FAILURE;

			/* frame 1: Pairwise == 1, Install == 0, Ack == 1, MIC == 0 */

			if ((h80211[z + 6] & 0x08) != 0 && (h80211[z + 6] & 0x40) == 0
				&& (h80211[z + 6] & 0x80) != 0
				&& (h80211[z + 5] & 0x01) == 0)
			{
				memcpy(st_cur->wpa.anonce, &h80211[z + 17], 32);
				st_cur->wpa.state = 1;
			}

			/* frame 2 or 4: Pairwise == 1, Install == 0, Ack == 0, MIC == 1 */

			if (z + 17 + 32 > caplen) return FAILURE;

			if ((h80211[z + 6] & 0x08) != 0 && (h80211[z + 6] & 0x40) == 0
				&& (h80211[z + 6] & 0x80) == 0
				&& (h80211[z + 5] & 0x01) != 0)
			{
				if (memcmp(&h80211[z + 17], ZERO, 32) != 0)
				{
					memcpy(st_cur->wpa.snonce, &h80211[z + 17], 32);
					st_cur->wpa.state |= 2;
				}
			}

			/* frame 3: Pairwise == 1, Install == 1, Ack == 1, MIC == 1 */

			if ((h80211[z + 6] & 0x08) != 0 && (h80211[z + 6] & 0x40) != 0
				&& (h80211[z + 6] & 0x80) != 0
				&& (h80211[z + 5] & 0x01) != 0)
			{
				st_cur->wpa.eapol_size
					= (h80211[z + 2] << 8) + h80211[z + 3] + 4;

				if (st_cur->wpa.eapol_size > sizeof(st_cur->wpa.eapol_size)
					|| caplen - z < st_cur->wpa.eapol_size)
				{
					// ignore packet trying to crash us
					st_cur->wpa.eapol_size = 0;
					return 0;
				}

				if (memcmp(&h80211[z + 17], ZERO, 32) != 0)
				{
					memcpy(st_cur->wpa.anonce, &h80211[z + 17], 32);
					st_cur->wpa.state |= 4;
				}

				memcpy(st_cur->wpa.keymic, &h80211[z + 81], 16);
				memcpy(st_cur->wpa.eapol, &h80211[z], st_cur->wpa.eapol_size);
				memset(st_cur->wpa.eapol + 81, 0, 16);
				st_cur->wpa.state |= 8;
				st_cur->wpa.keyver = h80211[z + 6] & 7;

				if (st_cur->wpa.state == 15)
				{
					memcpy(st_cur->wpa.stmac, st_cur->stmac, 6);

					if (G.f_ivs != NULL)
					{
						memset(&ivs2, '\x00', sizeof(struct ivs2_pkthdr));
						ivs2.flags = 0;
						ivs2.len = 0;

						ivs2.len = sizeof(struct WPA_hdsk);
						ivs2.flags |= IVS2_WPA;

						if (memcmp(G.prev_bssid, ap_cur->bssid, 6) != 0)
						{
							ivs2.flags |= IVS2_BSSID;
							ivs2.len += 6;
							memcpy(G.prev_bssid, ap_cur->bssid, 6);
						}

						if (fwrite(
								&ivs2, 1, sizeof(struct ivs2_pkthdr), G.f_ivs)
							!= (size_t) sizeof(struct ivs2_pkthdr))
						{
							perror("fwrite(IV header) failed");
							return (1);
						}

						if (ivs2.flags & IVS2_BSSID)
						{
							if (fwrite(ap_cur->bssid, 1, 6, G.f_ivs)
								!= (size_t) 6)
							{
								perror("fwrite(IV bssid) failed");
								return (1);
							}
							ivs2.len -= 6;
						}

						if (fwrite(&(st_cur->wpa),
								   1,
								   sizeof(struct WPA_hdsk),
								   G.f_ivs)
							!= (size_t) sizeof(struct WPA_hdsk))
						{
							perror("fwrite(IV wpa_hdsk) failed");
							return (1);
						}
						return WPA;
					}
				}
			}
		}
	}

	return (0);
}

int main(int argc, char * argv[])
{
	time_t tt;
	int n, unused, ret;
	FILE * f_in;
	unsigned long nbr;
	unsigned long nbivs;
	unsigned char * h80211;
	unsigned char bssid_cur[6];
	unsigned char bssid_prv[6];
	unsigned char buffer[65536];
	struct pcap_file_header pfh;
	struct pcap_pkthdr pkh;
	struct ivs2_filehdr fivs2;

	if (argc < 4)
	{
		usage(0);
		return (1);
	}

	if (strcmp(argv[1], "--merge") == 0)
	{
		return merge(argc, argv);
	}
	if (strcmp(argv[1], "--convert"))
	{
		usage(1);
		return (1);
	}

	// Check filenames are not empty
	if (argv[2][0] == 0)
	{
		printf("Invalid pcap file\n");
		return (1);
	}

	if (argv[3][0] == 0)
	{
		printf("Invalid output file\n");
		return (1);
	}

	memset(bssid_cur, 0, 6);
	memset(bssid_prv, 0, 6);

	/* check the input pcap file */

	printf("Opening %s\n", argv[2]);

	if ((f_in = fopen(argv[2], "rb")) == NULL)
	{
		perror("fopen failed");
		return (1);
	}

	n = sizeof(pfh);

	if (fread(&pfh, 1, n, f_in) != (size_t) n)
	{
		perror("fread(pcap file header) failed");
		fclose(f_in);
		return (1);
	}

	if (pfh.magic != TCPDUMP_MAGIC && pfh.magic != TCPDUMP_CIGAM)
	{
		printf("\"%s\" isn't a pcap file (expected "
			   "TCPDUMP_MAGIC).\n",
			   argv[2]);
		fclose(f_in);
		return (1);
	}

	if (pfh.magic == TCPDUMP_CIGAM) SWAP32(pfh.linktype);

	if (pfh.linktype != LINKTYPE_IEEE802_11
		&& pfh.linktype != LINKTYPE_PRISM_HEADER
		&& pfh.linktype != LINKTYPE_RADIOTAP_HDR
		&& pfh.linktype != LINKTYPE_PPI_HDR)
	{
		printf("\"%s\" isn't a regular 802.11 "
			   "(wireless) capture.\n",
			   argv[2]);
		fclose(f_in);
		return (1);
	}

	/* create the output ivs file */

	printf("Creating %s\n", argv[3]);

	if ((G.f_ivs = fopen(argv[3], "wb+")) == NULL)
	{
		perror("fopen failed");
		fclose(f_in);
		return (1);
	}

	fivs2.version = IVS2_VERSION;

	unused = fwrite(IVS2_MAGIC, 4, 1, G.f_ivs);
	unused = fwrite(&fivs2, sizeof(struct ivs2_filehdr), 1, G.f_ivs);

	nbr = 0;
	tt = time(NULL) - 1;

	nbivs = 0;

	while (1)
	{
		if (time(NULL) - tt > 0)
		{
			erase_line(0);
			printf("Read %lu packets...\r", nbr);
			fflush(stdout);
			tt = time(NULL);
		}

		/* read one packet */

		n = sizeof(pkh);

		if (fread(&pkh, 1, n, f_in) != (size_t) n) break;

		if (pfh.magic == TCPDUMP_CIGAM)
		{
			SWAP32(pkh.caplen);
			SWAP32(pkh.len);
		}

		n = pkh.caplen;

		if (n <= 0 || n > 65535)
		{
			printf("Corrupted file? Invalid packet length: %d.\n", n);
			return (1);
		}

		if (fread(buffer, 1, n, f_in) != (size_t) n) break;

		++nbr;

		h80211 = buffer;

		/* remove any prism/radiotap header */

		if (pfh.linktype == LINKTYPE_PRISM_HEADER)
		{
			if (h80211[7] == 0x40)
				n = 64;
			else
			{
				n = *(int *) (h80211 + 4);

				if (pfh.magic == TCPDUMP_CIGAM) SWAP32(n);
			}

			if (n < 8 || n >= (int) pkh.caplen) continue;

			h80211 += n;
			pkh.caplen -= n;
		}

		if (pfh.linktype == LINKTYPE_RADIOTAP_HDR)
		{
			n = *(unsigned short *) (h80211 + 2);

			if (n <= 0 || n >= (int) pkh.caplen) continue;

			h80211 += n;
			pkh.caplen -= n;
		}

		if (pfh.linktype == LINKTYPE_PPI_HDR)
		{
			/* Remove the PPI header */

			n = le16_to_cpu(*(unsigned short *) (h80211 + 2));

			if (n <= 0 || n >= (int) pkh.caplen) continue;

			/* for a while Kismet logged broken PPI headers */
			if (n == 24 && le16_to_cpu(*(unsigned short *) (h80211 + 8)) == 2)
				n = 32;

			if (n <= 0 || n >= (int) pkh.caplen) continue;

			h80211 += n;
			pkh.caplen -= n;
		}

		ret = dump_add_packet(h80211, pkh.caplen);

		if (ret == IVS) ++nbivs;
	}
	fclose(f_in);
	fclose(G.f_ivs);

	erase_line(2);
	printf("Read %lu packets.\n", nbr);

	if (nbivs > 0)
		printf("Written %lu IVs.\n", nbivs);
	else
	{
		remove(argv[3]);
		puts("No IVs written");
	}

	return (0);
}
